core/storage/merkle/polkadot_trie_db/polkadot_codec.cpp

```diff
@@ -0,0 +1,222 @@
+    }
+
+    if (last_nibble_0) {
+      out[nibbles] = low4nibbles(key[key.size() - 1]);
+    }
+
+    return out;
+  }
+
+  common::Hash256 PolkadotCodec::hash256(const common::Buffer &buf) const {
+    common::Hash256 out;
+
+    static constexpr uint32_t k256 = 32;
+    if (buf.size() < k256) {
+      std::copy(buf.begin(), buf.end(), out.begin());
+      return out;
+    }
+
+    blake2s(out.data(), k256, nullptr, 0, buf.toBytes(), buf.size());
+    return out;
+  }
+
+  outcome::result<common::Buffer> PolkadotCodec::nodeEncode(
+      const Node &node) const {
+    switch (node.getType()) {
+      case Node::Type::Branch:
+        return encodeBranch(dynamic_cast<const BranchNode &>(node));
+      case Node::Type::Leaf:
+        return encodeLeaf(dynamic_cast<const LeafNode &>(node));
+    }
+
+    return std::errc::invalid_argument;
+  }
+
+  outcome::result<common::Buffer> PolkadotCodec::getHeader(
+      const PolkadotNode &node) const {
+    if (node.key_nibbles.size() > 0xffffu) {
+      return Error::TOO_MANY_NIBBLES;
+    }
+
+    uint8_t head = 0;
+    // set bits 6..7
+    switch (node.getType()) {
+      case Node::Type::Branch:
+        // has value?
+        head = node.value.size() > 0 ? 0b11 : 0b10;
+        break;
+      case Node::Type::Leaf:
+        head = 0b01;
+        break;
+      default:
+        return Error::UNKNOWN_NODE_TYPE;
+    }
+    // FIXME: in the spec above two bits are most significant
+    //    head <<= 6u;  // head_{6..7} * 64
+
+    // set bits 0..5
+    if (node.key_nibbles.size() < 63u) {
+      // FIXME: in the spec it is
+      //   head |= uint8_t(node.key_nibbles.size());
+      head |= uint8_t(node.key_nibbles.size()) << 2u;
+      return Buffer{head};  // header contains 1 byte
+    }
+
+    // FIXME: in the spec it is += 63u; - set least 6 bits with 1
+    head += 0xfcu;
+
+    size_t l = node.key_nibbles.size() - 63u;
+    Buffer out(1u +             /// 1 byte head
+                   l / 0xffu +  /// number of 255 in l
+                   1u,          /// for last byte
+               0xffu            /// fill array with 255
+    );
+
+    // first byte is head
+    out[0] = head;
+    // last byte after while(l>255) l-=255;
+    out[out.size() - 1] = l % 0xffu;
+
+    return out;
+  }
+
+  outcome::result<common::Buffer> PolkadotCodec::encodeBranch(
+      const BranchNode &node) const {
+    // node header
+    OUTCOME_TRY(encoding, getHeader(node));
+
+    // key
+    encoding += nibblesToKey(node.key_nibbles);
+
+    // children bitmap
+    encoding += ushortToBytes(node.childrenBitmap());
+
+    // encode each child
+    for (const auto &child : node.children) {
+      if (child) {
+        OUTCOME_TRY(encChild, nodeEncode(*child));
+        encoding += encChild;
+      }
+    }
+
+    // scale encoded value
+    OUTCOME_TRY(encNodeValue, scale_->encode(node.value));
+    encoding += encNodeValue;
+
+    return outcome::success(std::move(encoding));
```

Comment:
Isn't `return encoding;` enough?

Reply:
clang-tidy suggests to write `std::move(encoding)`